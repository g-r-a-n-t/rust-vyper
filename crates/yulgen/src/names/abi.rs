use crate::types::{AbiDecodeLocation, AbiType};
use std::borrow::Borrow;
use yultsur::*;

/// Generates an ABI encoding function name for a given set of types.
pub fn encode(_types: &[AbiType]) -> yul::Identifier {
    let name = format!("abi_encode_{}", types(_types));

    identifier! { (name) }
}

pub fn decode_data(_types: &[AbiType], location: AbiDecodeLocation) -> yul::Identifier {
    let name = format!(
        "abi_decode_data_{}_{}",
        types(_types),
        decode_location(location)
    );

    identifier! { (name) }
}

pub fn decode_component(typ: &AbiType, location: AbiDecodeLocation) -> yul::Identifier {
    match typ {
        AbiType::Address => decode_component_address(location),
        AbiType::Bool => decode_component_bool(location),
        AbiType::Uint { size } => decode_component_uint(*size, location),
        AbiType::Int { size } => decode_component_int(*size, location),
        AbiType::StaticArray { inner, size } => {
            decode_component_static_array(inner, *size, location)
        }
        AbiType::Tuple { components: elems } => decode_component_tuple(elems, location),
        AbiType::String { max_size } => decode_component_string(*max_size, location),
        AbiType::Bytes { size } => decode_component_bytes(*size, location),
    }
}

pub fn decode_component_uint(size: usize, location: AbiDecodeLocation) -> yul::Identifier {
    let name = format!(
        "abi_decode_component_uint{}_{}",
        size * 8,
        decode_location(location)
    );

    identifier! { (name) }
}

pub fn decode_component_int(size: usize, location: AbiDecodeLocation) -> yul::Identifier {
    let name = format!(
        "abi_decode_component_int{}_{}",
        size * 8,
        decode_location(location)
    );

    identifier! { (name) }
}

pub fn decode_component_bool(location: AbiDecodeLocation) -> yul::Identifier {
    let name = format!("abi_decode_component_bool_{}", decode_location(location));

    identifier! { (name) }
}

pub fn decode_component_address(location: AbiDecodeLocation) -> yul::Identifier {
    let name = format!("abi_decode_component_address_{}", decode_location(location));

    identifier! { (name) }
}

pub fn decode_component_static_array(
    inner: &AbiType,
    size: usize,
    location: AbiDecodeLocation,
) -> yul::Identifier {
    let name = format!(
        "abi_decode_component_static_array_{}_{}_{}",
        size,
        typ(inner),
        decode_location(location)
    );

    identifier! { (name) }
}

pub fn decode_component_tuple(
    components: &[AbiType],
    location: AbiDecodeLocation,
) -> yul::Identifier {
    let name = format!(
        "abi_decode_component_tuple_{}_{}",
        types(components),
        decode_location(location)
    );

    identifier! { (name) }
}

pub fn decode_component_bytes(size: usize, location: AbiDecodeLocation) -> yul::Identifier {
    let name = format!(
        "abi_decode_component_bytes_{}_{}",
        size,
        decode_location(location)
    );

    identifier! { (name) }
}

pub fn decode_component_string(max_size: usize, location: AbiDecodeLocation) -> yul::Identifier {
    let name = format!(
        "abi_decode_component_string_{}_{}",
        max_size,
        decode_location(location)
    );

    identifier! { (name) }
}

pub fn vals(prefix: &str, n: usize) -> (Vec<yul::Identifier>, Vec<yul::Expression>) {
    (0..n)
        .into_iter()
        .map(|index| {
            let name = format!("{}_val_{}", prefix, index);
            (identifier! { (name) }, identifier_expression! { (name) })
        })
        .unzip()
}

fn typ(_typ: &AbiType) -> String {
    match _typ {
        AbiType::Uint { size } => format!("uint{}", size * 8),
        AbiType::Int { size } => format!("int{}", size * 8),
        AbiType::Bool => "bool".to_string(),
        AbiType::Address => "address".to_string(),
        AbiType::StaticArray { size, inner } => format!("array_{}_{}", size, typ(inner)),
        AbiType::Tuple { components } => format!("tuple_{}", types(components)),
        AbiType::String { max_size } => format!("string_{}", max_size),
        AbiType::Bytes { size } => format!("bytes_{}", size),
    }
}

pub fn types(types: &[AbiType]) -> String {
    types.iter().map(typ).collect::<Vec<_>>().join("_")
}

fn decode_location(location: AbiDecodeLocation) -> &'static str {
    match location {
        AbiDecodeLocation::Memory => "mem",
        AbiDecodeLocation::Calldata => "calldata",
    }
}
