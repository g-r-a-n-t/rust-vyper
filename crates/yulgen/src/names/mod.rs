use crate::names::abi as abi_names;
use crate::types::{to_abi_selector_names, AbiType};
use fe_abi::utils as abi_utils;
use fe_analyzer::namespace::types::{Integer, SafeNames};
use yultsur::*;

pub mod abi;

/// Generate a function name to perform checked addition
pub fn checked_add(size: &Integer) -> yul::Identifier {
    let size: &str = size.into();
    identifier! {(format!("checked_add_{}", size.to_lowercase()))}
}

/// Generate a function name to perform checked division
pub fn checked_div(size: &Integer) -> yul::Identifier {
    let size: &str = if size.is_signed() {
        size.into()
    } else {
        "unsigned"
    };
    identifier! {(format!("checked_div_{}", size.to_lowercase()))}
}

/// Generate a function name to perform checked modulo
pub fn checked_mod(size: &Integer) -> yul::Identifier {
    let sign: &str = if size.is_signed() {
        "signed"
    } else {
        "unsigned"
    };
    identifier! {(format!("checked_mod_{}", sign.to_lowercase()))}
}

/// Generate a function name to perform checked exponentiation
pub fn checked_exp(size: &Integer) -> yul::Identifier {
    let size: &str = size.into();
    identifier! {(format!("checked_exp_{}", size.to_lowercase()))}
}

/// Generate a function name to perform checked multiplication
pub fn checked_mul(size: &Integer) -> yul::Identifier {
    let size: &str = size.into();
    identifier! {(format!("checked_mul_{}", size.to_lowercase()))}
}

/// Generate a function name to perform checked subtraction
pub fn checked_sub(size: &Integer) -> yul::Identifier {
    let size: &str = if size.is_signed() {
        size.into()
    } else {
        "unsigned"
    };
    identifier! {(format!("checked_sub_{}", size.to_lowercase()))}
}

/// Generate a safe function name for a user defined function
pub fn func_name(name: &str) -> yul::Identifier {
    identifier! { (format!("$${}", name)) }
}

/// Generate a safe variable name for a user defined function
pub fn var_name(name: &str) -> yul::Identifier {
    identifier! { (format!("${}", name)) }
}

/// Generate a revert function name for the name `Error` and a given set of types
pub fn error_revert_name(types: &[AbiType]) -> yul::Identifier {
    revert_name("Error", types)
}

/// Generates a revert function name for a given name and types
pub fn revert_name(name: &str, types: &[AbiType]) -> yul::Identifier {
    let name = format!("revert_with_{}_{}", name, abi_names::types(types));

    identifier! { (name) }
}

/// Generates an external call function name for a given type and location.
pub fn contract_call(contract_name: &str, func_name: &str) -> yul::Identifier {
    let name = format!("{}_{}", contract_name, func_name);
    identifier! { (name) }
}

/// Generates a function name for to interact with a certain struct type
pub fn struct_function_name(struct_name: &str, func_name: &str) -> yul::Identifier {
    let name = format!("struct_{}_{}", struct_name, func_name);
    identifier! { (name) }
}

/// Generates a function name for creating a certain struct type
pub fn struct_new_call(struct_name: &str) -> yul::Identifier {
    struct_function_name(struct_name, "new")
}

/// Generates a function name for reading a named property of a certain struct
/// type
pub fn struct_getter_call(struct_name: &str, field_name: &str) -> yul::Identifier {
    struct_function_name(struct_name, &format!("get_{}_ptr", field_name))
}
