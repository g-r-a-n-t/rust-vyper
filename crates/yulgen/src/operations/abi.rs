use crate::names::abi as abi_names;
use crate::operations::data as data_operations;
use crate::types::{AbiType, AbiDecodeLocation};
use crate::utils::ceil_32;
use yultsur::*;

/// The size of an encoding known at compile-time.
///
/// The expressions should only be given literal values.
pub enum EncodingSize {
    Exact(yul::Expression),
    Bounded {
        min: yul::Expression,
        max: yul::Expression,
    },
}

/// Returns an expression that encodes the given values and returns a pointer to
/// the encoding.
pub fn encode(types: &[AbiType], vals: Vec<yul::Expression>) -> yul::Expression {
    let func_name = abi_names::encode(&types);
    expression! { [func_name]([vals...]) }
}

/// Returns an expression that gives size of the encoded values.
///
/// It will sum up the sizes known at compile-time with the sizes known during runtime.
pub fn encoding_size(types: &[AbiType], vals: Vec<yul::Expression>) -> yul::Expression {
    let mut head_size = 0;
    let mut known_data_size = 0;
    let mut unknown_data_size = vec![];

    let typed_vals = types.iter().zip(vals);

    for (typ, val) in typed_vals {
        head_size += typ.head_size();
        match typ {
            AbiType::String { .. } => {
                known_data_size += 32;
                unknown_data_size.push(expression! { ceil32((mload([val]))) })
            }
            AbiType::Bytes { size } => known_data_size += 32 + ceil_32(*size),
            _ => {}
        }
    }

    let static_size = literal_expression! { (head_size + known_data_size) };
    expression! { add([static_size], [data_operations::sum(unknown_data_size)]) }
}

/// Returns an expression that gives the size of the encoding's head.
pub fn encoding_head_size(types: &[AbiType]) -> yul::Expression {
    literal_expression! { (types.iter().map(|typ| typ.head_size()).sum::<usize>()) }
}

/// Returns the known-at-compile-time encoding size.
pub fn encoding_known_size(types: &[AbiType]) -> EncodingSize {
    let (min, max) = types.iter().fold((0, 0), |(mut min, mut max), typ| {
        min += typ.head_size();
        max += typ.head_size();

        match typ {
            AbiType::String { max_size } => {
                min += 32;
                max += ceil_32(*max_size) + 32;
            }
            AbiType::Bytes { size } => {
                let size = ceil_32(*size) + 32;
                min += size;
                max += size;
            }
            _ => {}
        }

        (min, max)
    });

    if min == max {
        EncodingSize::Exact(literal_expression! { (min) })
    } else {
        EncodingSize::Bounded {
            min: literal_expression! { (min) },
            max: literal_expression! { (max) },
        }
    }
}

/// Decode a segment of memory and return each decoded component as separate values.
pub fn decode_data(
    types: &[AbiType],
    head_start: yul::Expression,
    data_end: yul::Expression,
    location: AbiDecodeLocation,
) -> yul::Expression {
    let func_name = abi_names::decode_data(types, location);
    expression! { [func_name]([head_start], [data_end]) }
}

/// Decode a single component.
pub fn decode_component(
    typ: &AbiType,
    head_start: yul::Expression,
    offset: yul::Expression,
    location: AbiDecodeLocation,
) -> yul::Expression {
    let func_name = abi_names::decode_component(typ, location);
    expression! { [func_name]([head_start], [offset]) }
}

/// Unpack each value into a newly allocated segment of memory.
pub fn unpack(
    ptr: yul::Expression,
    array_size: yul::Expression,
    inner_data_size: yul::Expression,
) -> yul::Statement {
    statement! { abi_unpack([ptr], [array_size], [inner_data_size]) }
}
