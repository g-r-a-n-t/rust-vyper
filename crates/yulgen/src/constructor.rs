use crate::names::abi as abi_names;
use crate::operations::abi as abi_operations;
use crate::types::{to_abi_types, AbiDecodeLocation};
use fe_analyzer::namespace::types::{FixedSize};
use yultsur::*;

/// Builds a constructor for a contract with no init function.
///
/// The contract is simply deployed by loading the code into memory and
/// returning it.
pub fn build() -> yul::Code {
    // we get the deployment statements and wrap them in a code block
    let deployment = deployment();
    code! { [deployment...] }
}

/// Builds a constructor for a contract with an init function.
///
/// We include the entire contact runtime inside of the constructor (without the
/// ABI dispatcher), run the init function, and return the contract code.
pub fn build_with_init(
    contract_name: &str,
    init_func: yul::Statement,
    init_params: Vec<FixedSize>,
    runtime: Vec<yul::Statement>,
) -> yul::Code {
    // Generate names for our constructor parameters.
    let (param_idents, param_exprs) = abi_names::vals("init", init_params.len());

    // Decode the parameters, if any are given.
    let maybe_decode_params = if init_params.is_empty() {
        statements! {}
    } else {
        let decode_expr = abi_operations::decode_data(
            &to_abi_types(&init_params),
            expression! { params_start_mem },
            expression! { params_end_mem },
            AbiDecodeLocation::Memory,
        );

        statements! { (let [param_idents...] := [decode_expr]) }
    };

    // init function name after it is mapped.
    let init_func_name = identifier! { ("$$__init__") };

    let contract_name = literal_expression! { (format!("\"{}\"", contract_name)) };

    let deployment = deployment();

    // Build a constructor that runs a user defined init function. Parameters for
    // init functions are appended to the end of the initialization code.
    //
    // The start of the init parameters in code is stored in `params_start_code`.
    // The parameters are immediately copied from this location into memory at
    // `mem_start`. From there, parameters are decoded and passed into the
    // init function.
    code! {
        // add init function to the scope
        [init_func]

        // add the entire contract runtime
        [runtime...]

        // copy the encoded parameters to memory
        (let params_start_code := datasize([contract_name]))
        (let params_end_code := codesize())
        (let params_size := sub(params_end_code, params_start_code))
        (let params_start_mem := alloc(params_size))
        (let params_end_mem := add(params_start_mem, params_size))
        (codecopy(params_start_mem, params_start_code, params_size))

        // decode the parameters from memory
        [maybe_decode_params...]

        // call the init function defined above
        (pop(([init_func_name]([param_exprs...]))))

        // deploy the contract
        [deployment...]
    }
}

/// Copies contract data to memory and returns it.
fn deployment() -> Vec<yul::Statement> {
    statements! {
        (let size := datasize("runtime"))
        (datacopy(0, (dataoffset("runtime")), size))
        (return(0, size))
    }
}
