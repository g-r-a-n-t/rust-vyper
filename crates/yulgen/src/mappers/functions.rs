use crate::constants::PANIC_FAILED_ASSERTION;
use crate::mappers::{assignments, declarations, expressions};
use crate::names;
use crate::operations::abi as abi_operations;
use crate::operations::data as data_operations;
use crate::types::{to_abi_types, AbiType, EvmSized};
use crate::Context;
use fe_analyzer::context::ExpressionAttributes;
use fe_analyzer::namespace::types::{FixedSize, Type};
use fe_parser::ast as fe;
use fe_parser::node::Node;
use yultsur::*;

pub fn multiple_func_stmt(
    context: &mut Context,
    statements: &[Node<fe::FuncStmt>],
) -> Vec<yul::Statement> {
    statements
        .iter()
        .map(|statement| func_stmt(context, statement))
        .collect()
}

/// Builds a Yul function definition from a Fe function definition.
pub fn func_def(context: &mut Context, def: &Node<fe::Function>) -> yul::Statement {
    let fe::Function {
        name, args, body, ..
    } = &def.kind;
    let function_name = names::func_name(&name.kind);
    let param_names = args.iter().map(|arg| func_def_arg(arg)).collect::<Vec<_>>();
    let function_statements = multiple_func_stmt(context, body);

    // all user-defined functions are given a return value during lowering
    function_definition! {
        function [function_name]([param_names...]) -> return_val {
            [function_statements...]
        }
    }
}

fn func_def_arg(arg: &Node<fe::FunctionArg>) -> yul::Identifier {
    let name = &arg.kind.name.kind;

    names::var_name(name)
}

fn func_stmt(context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    match &stmt.kind {
        fe::FuncStmt::Return { .. } => func_return(context, stmt),
        fe::FuncStmt::VarDecl { .. } => declarations::var_decl(context, stmt),
        fe::FuncStmt::Assign { .. } => assignments::assign(context, stmt),
        fe::FuncStmt::Emit { .. } => emit(context, stmt),
        fe::FuncStmt::AugAssign { .. } => panic!("AugAssign should be lowered"),
        fe::FuncStmt::For { .. } => for_loop(context, stmt),
        fe::FuncStmt::While { .. } => while_loop(context, stmt),
        fe::FuncStmt::If { .. } => if_statement(context, stmt),
        fe::FuncStmt::Assert { .. } => assert(context, stmt),
        fe::FuncStmt::Expr { .. } => expr(context, stmt),
        fe::FuncStmt::Pass => statement! { pop(0) },
        fe::FuncStmt::Break => break_statement(context, stmt),
        fe::FuncStmt::Continue => continue_statement(context, stmt),
        fe::FuncStmt::Revert { .. } => revert(context, stmt),
    }
}

fn for_loop(context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    if let fe::FuncStmt::For { target, iter, body } = &stmt.kind {
        let iterator = expressions::expr(context, iter);
        let target_var = names::var_name(&target.kind);
        let yul_body = multiple_func_stmt(context, body);
        return if let Some(ExpressionAttributes {
            typ: Type::Array(array),
            ..
        }) = context.analysis.get_expression(iter)
        {
            let size = literal_expression! { (array.size) };
            let inner_size = literal_expression! { (array.inner.size()) };
            block_statement! {
                (for {(let i := 0)} (lt(i, [size])) {(i := add(i, 1))}
                {
                    // Below yul statement to load values from memory to `target_var`.
                    (let [target_var] := [expression! { mloadn([expression! { add([iterator], (mul(i, [inner_size.clone()]))) }], [inner_size]) }])
                    [yul_body...]
                })
            }
        } else {
            panic!("missing iter expression")
        };
    }
    unreachable!()
}

fn if_statement(context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    if let fe::FuncStmt::If {
        test,
        body,
        or_else,
    } = &stmt.kind
    {
        let yul_test = expressions::expr(context, &test);
        let yul_body = multiple_func_stmt(context, body);
        let yul_or_else = multiple_func_stmt(context, or_else);

        return switch! {
            switch ([yul_test])
            (case 1 {[yul_body...]})
            (case 0 {[yul_or_else...]})
        };
    }

    unreachable!()
}

fn expr(context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    if let fe::FuncStmt::Expr { value } = &stmt.kind {
        let expr = expressions::expr(context, value);

        statement! { pop([expr])}
    } else {
        unreachable!()
    }
}

fn revert(context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    if let fe::FuncStmt::Revert { error } = &stmt.kind {
        if let Some(error_expr) = error {
            let error_attributes = context
                .analysis
                .get_expression(error_expr)
                .expect("missing expression");

            if let Type::Struct(val) = &error_attributes.typ {
                context.revert_errors.insert(val.clone());

                let revert_data = expressions::expr(context, error_expr);
                let size =
                    abi_operations::encoding_size(&[AbiType::from(val)], vec![revert_data.clone()]);
                let revert_fn = names::revert_name(&val.name, &to_abi_types(&val.get_field_types()));

                return statement! {
                    ([revert_fn]([revert_data], [size]))
                };
            }
        }

        return statement! { revert(0, 0) };
    }

    unreachable!()
}

fn emit(context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    if let fe::FuncStmt::Emit { args, .. } = &stmt.kind {
        let event_values = args
            .kind
            .iter()
            .map(|arg| expressions::expr(context, &arg.kind.value))
            .collect();

        if let Some(event) = context.analysis.get_emit(stmt) {
            return data_operations::emit_event(event.to_owned(), event_values);
        }

        panic!("missing event definition");
    }

    unreachable!()
}

fn assert(context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    if let fe::FuncStmt::Assert { test, msg } = &stmt.kind {
        let test = expressions::expr(context, test);
        return match msg {
            Some(val) => {
                let msg = expressions::expr(context, val);
                let msg_attributes = context
                    .analysis
                    .get_expression(val)
                    .expect("missing expression");

                if let Type::String(str) = &msg_attributes.typ {
                    let size =
                        abi_operations::encoding_size(&[AbiType::from(str)], vec![msg.clone()]);
                    let fixed_size = FixedSize::String(str.clone());
                    context.assert_strings.insert(str.clone());
                    let revert_fn = names::error_revert_name(&[AbiType::from(&fixed_size)]);

                    return statement! {
                        if (iszero([test])) {
                            ([revert_fn]([msg], [size]))
                        }
                    };
                }
                unreachable!()
            }
            None => {
                statement! { if (iszero([test])) { (revert_with_panic([literal_expression! {(PANIC_FAILED_ASSERTION)}])) } }
            }
        };
    }

    unreachable!()
}

fn break_statement(_context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    if let fe::FuncStmt::Break {} = &stmt.kind {
        return statement! { break };
    }

    unreachable!()
}

fn continue_statement(_context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    if let fe::FuncStmt::Continue {} = &stmt.kind {
        return statement! { continue };
    }

    unreachable!()
}

fn func_return(context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    if let fe::FuncStmt::Return { value } = &stmt.kind {
        let value = value
            .as_ref()
            .expect("valueless return made it to Yul codegen");
        let value = expressions::expr(context, value);

        block_statement! {
            (return_val := [value])
            (leave)
        }
    } else {
        unreachable!()
    }
}

fn while_loop(context: &mut Context, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
    if let fe::FuncStmt::While { test, body } = &stmt.kind {
        let test = expressions::expr(context, test);
        let yul_body = multiple_func_stmt(context, body);

        return block_statement! {
            (for {} ([test]) {}
            {
                [yul_body...]
            })
        };
    }

    unreachable!()
}
