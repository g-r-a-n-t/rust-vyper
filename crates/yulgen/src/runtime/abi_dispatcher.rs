use crate::names::abi as abi_names;
use crate::operations::abi as abi_operations;
use crate::types::{to_abi_types, AbiDecodeLocation, AbiType};
use fe_abi::utils as abi_utils;
use fe_analyzer::context::FunctionAttributes;
use fe_analyzer::namespace::types::FixedSize;
use yultsur::*;

/// Builds a switch statement that dispatches calls to the contract.
pub fn dispatcher(attributes: &[FunctionAttributes]) -> yul::Statement {
    let arms = attributes
        .iter()
        .map(|arm| dispatch_arm(arm.to_owned()))
        .collect::<Vec<_>>();

    if arms.is_empty() {
        statement! { return(0, 0) }
    } else {
        switch! {
            switch (cloadn(0, 4))
            [arms...]
            (default { (return(0, 0)) })
        }
    }
}

fn dispatch_arm(attributes: FunctionAttributes) -> yul::Case {
    let selector = selector(&attributes.name, &to_abi_types(&attributes.param_types()));

    let (param_idents, param_exprs) = abi_names::vals("call", attributes.params.len());

    // If there are no params, we create an empty vector.
    let maybe_decode_params = if attributes.params.is_empty() {
        statements! {}
    } else {
        let decode_expr = abi_operations::decode_data(
            &to_abi_types(&attributes.param_types()),
            expression! { 4 },
            expression! { calldatasize() },
            AbiDecodeLocation::Calldata,
        );
        statements! { (let [param_idents...] := [decode_expr]) }
    };

    // If the function returns a unit value, we call the function and return
    // nothing. Otherwise, we encode the value and return it.
    let call_and_maybe_encode_return = {
        let name = identifier! { (format!("$${}", attributes.name)) };
        let call = expression! { [name]([param_exprs...]) };
        if attributes.return_type.is_unit() {
            statements! {
                (pop([call]))
                (return(0, 0))
            }
        } else {
            let encode_expr = abi_operations::encode(
                &[AbiType::from(&attributes.return_type)],
                expressions! { return_val },
            );
            let encoding_size = abi_operations::encoding_size(
                &[AbiType::from(&attributes.return_type)],
                expressions! { return_val },
            );
            statements! {
                (let return_val := [call])
                (let encoding_start := [encode_expr])
                (let encoding_size := [encoding_size])
                (return(encoding_start, encoding_size))
            }
        }
    };

    case! {
        case [selector] {
            [maybe_decode_params...]
            [call_and_maybe_encode_return...]
        }
    }
}

fn selector(name: &str, params: &[AbiType]) -> yul::Literal {
    let params = params
        .iter()
        .map(|param| param.selector_name())
        .collect::<Vec<String>>();

    literal! { (abi_utils::func_selector(name, &params)) }
}

#[cfg(test)]
mod tests {
    use crate::runtime::abi_dispatcher::selector;
    use crate::types::AbiType;
    use fe_analyzer::namespace::types::{FixedSize, U256};

    #[test]
    fn test_selector_literal_basic() {
        assert_eq!(selector("foo", &[]).to_string(), String::from("0xc2985578"),)
    }

    #[test]
    fn test_selector_literal() {
        assert_eq!(
            selector("bar", &[AbiType::Uint { size: 32 }]).to_string(),
            String::from("0x0423a132"),
        )
    }
}
