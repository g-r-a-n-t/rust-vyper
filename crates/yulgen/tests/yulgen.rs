use fe_analyzer::context::FunctionAttributes;
use fe_analyzer::namespace::events::EventDef;
use fe_analyzer::namespace::types::{
    Base, FeString, FixedSize, Integer, Struct, U256,
};
use fe_yulgen::constructor;
use fe_yulgen::names::abi as abi_names;
use fe_yulgen::operations::{abi as abi_operations, data as data_operations};
use fe_yulgen::runtime::abi_dispatcher;
use fe_yulgen::runtime::functions::{
    abi as abi_functions, revert as revert_functions, structs as structs_functions,
};
use fe_yulgen::types::{AbiType, AbiDecodeLocation};
use insta::assert_display_snapshot;
use wasm_bindgen_test::wasm_bindgen_test;
use yultsur::*;

macro_rules! test_yulgen {
    ($name:ident, $code:expr) => {
        #[test]
        #[wasm_bindgen_test]
        fn $name() {
            if cfg!(target_arch = "wasm32") {
                fe_common::assert_snapshot_wasm!(
                    concat!("snapshots/yulgen__", stringify!($name), ".snap"),
                    $code.to_string()
                );
            } else {
                assert_display_snapshot!($code);
            }
        }
    };
}

// constructor
test_yulgen! { constructor_no_init,  constructor::build() }

fn function_attributes() -> Vec<FunctionAttributes> {
    vec![
        FunctionAttributes {
            is_public: true,
            name: "hello_world".to_string(),
            params: vec![],
            return_type: FixedSize::String(FeString { max_size: 42 }),
        },
        FunctionAttributes {
            is_public: true,
            name: "add".to_string(),
            params: vec![
                (
                    "a".to_string(),
                    FixedSize::Base(Base::Numeric(Integer::U256)),
                ),
                (
                    "b".to_string(),
                    FixedSize::Base(Base::Numeric(Integer::U256)),
                ),
            ],
            return_type: FixedSize::Base(Base::Numeric(Integer::U256)),
        },
    ]
}

// ABI dispatcher
test_yulgen! { abi_dispatcher,  abi_dispatcher::dispatcher(&function_attributes()) }

// ABI encoding functions
test_yulgen! {
    abi_encode_u256_address_function,
    abi_functions::encode(vec![AbiType::Uint { size: 32 }, AbiType::Address])
}

// ABI decoding functions
test_yulgen! {
    abi_decode_data_address_bool_mem_function,
    abi_functions::decode_data(&[AbiType::Bool, AbiType::Address], AbiDecodeLocation::Memory)
}
test_yulgen! {
    abi_decode_data_u256_bytes_string_bool_address_bytes_calldata_function,
    abi_functions::decode_data(&[
        AbiType::Uint { size: 32 },
        AbiType::Bytes { size: 100 },
        AbiType::String { max_size: 42 },
        AbiType::Bool,
        AbiType::Address,
        AbiType::Bytes { size: 100 },
    ], AbiDecodeLocation::Calldata)
}
test_yulgen! {
    abi_decode_component_uint256_mem_function,
    abi_functions::decode_component_uint(32, AbiDecodeLocation::Memory)
}
test_yulgen! {
    abi_decode_component_int16_calldata_function,
    abi_functions::decode_component_int(2, AbiDecodeLocation::Calldata)
}
test_yulgen! {
    abi_decode_component_bool_calldata_function,
    abi_functions::decode_component_bool(AbiDecodeLocation::Calldata)
}
test_yulgen! {
    abi_decode_component_address_mem_function,
    abi_functions::decode_component_bool(AbiDecodeLocation::Memory)
}
test_yulgen! {
    abi_decode_component_static_array_address_calldata_function,
    abi_functions::decode_component_static_array(&AbiType::Address, 42, AbiDecodeLocation::Calldata)
}
test_yulgen! {
    abi_decode_component_tuple_u256_address_mem_function,
    abi_functions::decode_component_tuple(
        &[AbiType::Uint { size: 32 }, AbiType::Address],
        AbiDecodeLocation::Memory
    )
}
test_yulgen! {
    abi_decode_component_bytes_26_mem_function,
    abi_functions::decode_component_bytes(26, AbiDecodeLocation::Memory)
}
test_yulgen! {
    abi_decode_component_string_26_calldata_function,
    abi_functions::decode_component_bytes(26, AbiDecodeLocation::Calldata)
}

fn struct_bool_bool() -> Struct {
    let mut val = Struct::new("Foo");
    val.add_field("bar", &FixedSize::bool()).unwrap();
    val.add_field("bar2", &FixedSize::bool()).unwrap();
    val
}

// struct functions
test_yulgen! {
    struct_empty_function,
    structs_functions::generate_new_fn(&Struct::new("Foo"))
}
test_yulgen! {
    struct_new_gen_function,
    structs_functions::generate_new_fn(&struct_bool_bool())
}
test_yulgen! {
    struct_getter_gen_bar_function,
    structs_functions::generate_get_fn(&struct_bool_bool(), &struct_bool_bool().fields[0].0)
}
test_yulgen! {
    struct_getter_gen_bar2_function,
    structs_functions::generate_get_fn(&struct_bool_bool(), &struct_bool_bool().fields[1].0)
}

fn event_no_indexed() -> EventDef {
    EventDef::new(
        "MyEvent",
        vec![
            ("my_u256".to_string(), FixedSize::Base(U256)),
            ("my_addr".to_string(), FixedSize::Base(Base::Address)),
        ],
        vec![],
    )
}

fn event_one_indexed() -> EventDef {
    EventDef::new(
        "MyEvent",
        vec![
            ("my_u256".to_string(), FixedSize::Base(U256)),
            ("my_addr".to_string(), FixedSize::Base(Base::Address)),
        ],
        vec![0],
    )
}

// data operations
test_yulgen! {
    emit_event_no_indexed_operation,
    data_operations::emit_event(event_no_indexed(), vec![expression! { 26 }, expression! { 0x42 }])
}
test_yulgen! {
    emit_event_one_indexed_operation,
    data_operations::emit_event(event_one_indexed(), vec![expression! { 26 }, expression! { 0x42 }])
}
test_yulgen! {
    sum_operation,
    data_operations::sum(expressions! { 42 26 99 })
}

// ABI operations
test_yulgen! {
    encode_u256_operation,
    abi_operations::encode(&[AbiType::Uint { size: 32 }], vec![expression! { 42 }])
}
test_yulgen! {
    encode_size_u256_operation,
    abi_operations::encoding_size(&[AbiType::Uint { size: 32 }], vec![expression! { 42 }])
}
test_yulgen! {
    decode_string_calldata_operation,
    abi_operations::decode_data(
        &[AbiType::String { max_size: 26 }],
        expression! { 42 },
        expression! { 10 },
        AbiDecodeLocation::Calldata
    )
}

// ABI names
test_yulgen! {
    encode_name,
    abi_names::encode(&[
        AbiType::Uint { size: 32 },
        AbiType::Bytes { size: 100 }
    ])
}
test_yulgen! {
    decode_data_name_u256_calldata_name,
    abi_names::decode_data(&[AbiType::Uint { size: 32 }], AbiDecodeLocation::Calldata)
}
test_yulgen! {
    decode_data_name_string_mem_name,
    abi_names::decode_data(&[AbiType::String { max_size: 42 }], AbiDecodeLocation::Memory)
}

// revert functions
test_yulgen! {
    revert_string_error,
    revert_functions::generate_revert_fn_for_assert(&[AbiType::String { max_size: 3 }])
}
