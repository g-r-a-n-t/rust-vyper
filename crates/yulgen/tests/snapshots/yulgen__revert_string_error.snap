---
source: crates/yulgen/tests/yulgen.rs
expression: "revert_functions::generate_revert_fn_for_assert(&[AbiType::String{max_size:\n                                                                      3,}])"

---
function revert_with_Error_string_3(params_val_0) {
    let ptr := alloc_mstoren(0x08c379a0, 4)
    pop(abi_encode_string_3(params_val_0))
    revert(ptr, add(4, add(64, ceil32(mload(params_val_0)))))
}
